### ISSUE TITLE
Papyrus Compare unexplained model differences

### ISSUE THREAD

**Issue by eposse** [ZeligsoftDev/CX4CBDDS#156]
# **Issue and tracking information**

### Developer's Actual time spent on fix (hours)  

# **Issue reporter to provide a detailed description of the issue in the space below**

Note: This may not be a bug. It is simply that the results of a model diff with Papyrus Compare may not be completely clear.

When using Papyrus Compare to compare models, you see several differences which are not clear. For example, adding a part to a component results in several changes throughout the model in addition to just adding the part.

What is not clear is whether the changes seen are to be expected or not.

#### Steps to reproduce

(This assumes that you have the CX4CBDDS git repository added in the git perspective and the 'papyrus' branch checked-out).

1. Import the BasicPubSub_new project from the CX4CBDDS git repo 'examples' folder into the workspace
1. Open the BasicPubSub_new model (BasicPubSub/ModelFiles/BasicPubSubPapyrus)
1. Open the assembly diagram (In the Model Explorer navigate to /BasicPubSub/assemblies/BasicPubSub_asm/BasicPubSub_asm and double click on the "NewCompositeStructureDiagram" item to open it)
1. Add a new subscriber part (In the Model Explorer navigate to BasicPubSub/components/SNA_Examples/BasicSubscriber_comp, drag the BasicSubscriber_comp interface and drop it into the diagram).
1. Save the model.
1. Open the Git perspective (Window > Perspective > Open Perspective > Other... > Git)
1. Go to the "Git Staging" view.
1. Stage the /BasicPubSub/ModelFiles/BasicPubSubPapyrus.notation and the BasicPubSub/ModelFiles/BasicPubSubPapyrus.uml files. You may need to click the "Refresh" button in the top-right of the "Git Staging" view.
1. Go back to the Papyrus perspective (Window > Perspective > Open Perspective > Other... > Papyrus)
1. Right click on /BasicPubSub/ModelFiles/BasicPubSubPapyrus and select "Compare With > Index)
If prompted to switch to the Synchronization perspective, click on [Yes] (and optionally check the "Remember my decision" checkbox)
1. In the Synchronization view on the left, unfold BasicPubSub/ModelFiles, right-click on BasicPubSubPapyrus.notation and select "Open in Compare Editor".

#### Expected result

We expect to see the new part addition in both 

1. the structure of the component in which it is being added (/BasicPubSub/assemblies/BasicPubSub_asm/BasicPubSub_asm/basicSubscriber_comp1 : BasicSubscriber_comp ownedAttribute add]) and 
1. the corresponding diagram (/BasicPubSub/assemblies/BasicPubSub_asm/BasicPubSub_asm/Diagram NewCompositeStructureDiagram BasicPubSuc_asm/Shape BasicPubSub_asm/BasicCompartment BasicPubSub_asm/Shape basicSubscriber_comp1 : BasicSubscriber_comp [children add]).

#### Actual result

In addition to the the change in structure and diagram, several more differences are seen:

1. Additions to the Deployment plan: (/BasicPubSub/assemblies/Deployment/descriptors/BasicPubSub_asm/BasicPubSub_asm : BasicPubSub_asm/basicSubscriber_comp1 : BasicSubscriber_comp [nestedPart add] and /BasicPubSub/assemblies/Deployment/descriptors/BasicPubSub_asm/basicSubscriber_comp1 : BasicSubscriber_comp/basicSubscriber_comp1 : BasicSubscriber_comp [ownedAttribute add])
1. Additions (two directions) in the ports package: (/BasicPubSub/ports/SNA_Examples/BasicPubSub/TestData_conn/Reader/query : QueryFilter/InternalError, change in both directions; similarly for DDS_Read, DDS_Get, DDS_Listen and DDS_StateListen under TestData_conn).

**Comment by eposse** [comment]
The following may be a different issue, but it falls under "unexplained differences": the model diff shown for /BasicPubSub/ports/SNA_Examples/BasicPubSub/TestData_conn/Reader/query : QueryFilter shows a change in the `InternalError` exception [setraises add] immediately below the `query`attribute but inspecting the .uml source below shows that the the exception is raised by an operation (see the `raiseException` element below). In fact, by all operations under `Reader`. Why is Papyrus Compare putting it under the attribute?

```xml
            <packagedElement xmi:type="uml:Interface" xmi:id="_8LuWoLIjEeCxJ7b7h4plXw" name="Reader">
              <ownedAttribute xmi:type="uml:Property" xmi:id="_8MrY4LIjEeCxJ7b7h4plXw" name="query">
                <type xmi:type="uml:DataType" href="pathmap://DDS4CCM_DCPS_CONNECTOR_LIBRARIES/DDS_DCPS.uml#_w_ULgGRNEd-Wc9hRqhM12g"/>
              </ownedAttribute>
              <ownedOperation xmi:type="uml:Operation" xmi:id="_8LxZ8LIjEeCxJ7b7h4plXw" name="read_last">
                <ownedParameter xmi:type="uml:Parameter" xmi:id="_8LzPILIjEeCxJ7b7h4plXw" name="data" direction="inout">
                  <type xmi:type="uml:DataType" href="pathmap://DDS4CCM_DCPS_CONNECTOR_LIBRARIES/DDS_DCPS.uml#_l3JJgWTXEd-fwvmMvf-phA"/>
                </ownedParameter>
                <ownedParameter xmi:type="uml:Parameter" xmi:id="_8L2ScLIjEeCxJ7b7h4plXw" name="infos" direction="inout">
                  <type xmi:type="uml:DataType" href="pathmap://DDS4CCM_DCPS_CONNECTOR_LIBRARIES/DDS_DCPS.uml#_pjyOEGRNEd-Wc9hRqhM12g"/>
                </ownedParameter>
                <ownedParameter xmi:type="uml:Parameter" xmi:id="_8L4usLIjEeCxJ7b7h4plXw" name="Parameter1" direction="return">
                  <type xmi:type="uml:DataType" href="pathmap://DDS4CCM_LIBRARIES/IDLPrimitives.uml#_Iz6ZsExREd2INdt7Ch5VKg"/>
                </ownedParameter>
                <raisedException xmi:type="uml:DataType" href="pathmap://DDS4CCM_DCPS_CONNECTOR_LIBRARIES/DDS_DCPS.uml#_XRS4sGROEd-Wc9hRqhM12g"/>
...
```

**Comment by eposse** [comment]
The following attachment shows the textual diff between the BasicPubSubPapyrus.uml file before and after adding the part. The diff doesn't show the changes under /BasicPubSub/ports/SNA_Examples. Maybe it's a Papyrus Compare bug?

[adding-part.diff.txt](https://github.com/ZeligsoftDev/CX4CBDDS/files/4839583/adding-part.diff.txt)

**Comment by eposse** [comment]
The additions to the Deployment plan seem to be expected after all. Young-Soo said that adding a part would automatically make updates to the Deployment plan.

However, the differences under /BasicPubSub/ports/SNA_Examples/BasicPubSub/TestData_conn are incorrect, which is why I'm changing the label of this issue to a bug.

The problem here is that all the elements marked with differences under TestData_conn are not actually in the model but in one of the model libraries, namely `DDS_DCSP.uml`. See, for example, the type of the `query` attribute of `Reader` in the snippet from `BasicPubSubPapyrus.uml` in the comment above. It includes a cross-reference `pathmap://DDS4CCM_DCPS_CONNECTOR_LIBRARIES/DDS_DCPS.uml#_w_ULgGRNEd-Wc9hRqhM12g`, and furthermore, it has the `CORBAAttribute` stereotype applied, which also includes a cross-reference to the `DDS_DCSP.uml` model library:

```xml
  <dds4ccm:CORBAAttribute xmi:id="_8Mr_8LIjEeCxJ7b7h4plXw" base_NamedElement="_8MrY4LIjEeCxJ7b7h4plXw" base_Property="_8MrY4LIjEeCxJ7b7h4plXw">
    <setraises xmi:type="dds4ccm:CORBAException" href="pathmap://DDS4CCM_DCPS_CONNECTOR_LIBRARIES/DDS_DCPS.uml#_XRXKIGROEd-Wc9hRqhM12g"/>
  </dds4ccm:CORBAAttribute>
```

The screenshot attached shows the first of those unexpected erroneous differences.

<img width="1036" alt="unexpected_diff" src="https://user-images.githubusercontent.com/3011155/87093004-74027100-c20a-11ea-928b-f6930cf96740.png">

**Comment by eposse** [comment]
Note that it is in fact rather strange that the .uml file contains the `Reader`, `DDS_Read`, `DDS_Get`, `DDS_Listen` and `DDS_StateListen` elements, as these are also elements in the `DDS_DCSP.uml` model library. More precisely, the `BasicPubSubPapyrus.uml` file contains a (shallow) copy of these elements.It would appear that these elements where _copied_ from the library, rather than imported. This may explain the difference.

For example, the following is a snippet from `BasicPubSubPapyrus.uml`:

```xml
            <packagedElement xmi:type="uml:Interface" xmi:id="_8LuWoLIjEeCxJ7b7h4plXw" name="Reader">
              <ownedAttribute xmi:type="uml:Property" xmi:id="_8MrY4LIjEeCxJ7b7h4plXw" name="query">
                <type xmi:type="uml:DataType" href="pathmap://DDS4CCM_DCPS_CONNECTOR_LIBRARIES/DDS_DCPS.uml#_w_ULgGRNEd-Wc9hRqhM12g"/>
              </ownedAttribute>
...
```

and the following is from `DDS_DCSP.uml`:

```xml
        <packagedElement xmi:type="uml:Interface" xmi:id="_AaRikGTbEd-fwvmMvf-phA" name="Reader">
          <ownedAttribute xmi:type="uml:Property" xmi:id="__-74IGTfEd-fwvmMvf-phA" name="query" visibility="public" type="_w_ULgGRNEd-Wc9hRqhM12g"/>
...
```

Similarly, the stereotype applied to the `query` element is the following in `BasicPubSubPapyrus.uml`:

```xml
  <dds4ccm:CORBAAttribute xmi:id="_8Mr_8LIjEeCxJ7b7h4plXw" base_NamedElement="_8MrY4LIjEeCxJ7b7h4plXw" base_Property="_8MrY4LIjEeCxJ7b7h4plXw">
    <setraises xmi:type="dds4ccm:CORBAException" href="pathmap://DDS4CCM_DCPS_CONNECTOR_LIBRARIES/DDS_DCPS.uml#_XRXKIGROEd-Wc9hRqhM12g"/>
  </dds4ccm:CORBAAttribute>
```
and the same stereotype in `DDS_DCSP.uml`:

```xml
  <dds4ccm:CORBAAttribute xmi:id="__-_igGTfEd-fwvmMvf-phA" base_NamedElement="__-74IGTfEd-fwvmMvf-phA" setraises="_XRXKIGROEd-Wc9hRqhM12g" base_Property="__-74IGTfEd-fwvmMvf-phA"/>
...
```

Note that in `BasicPubSubPapyrus.uml` the `setraises` clause is different because it is a cross-reference  to the `CORBAException` element with id `_XRXKIGROEd-Wc9hRqhM12g` in `DDS_DCSP.uml` model, while in the later, `setraises` contains a direct reference to the local `CORBAException` element with the same id `_XRXKIGROEd-Wc9hRqhM12g` which is in the same file:

```xml
  <dds4ccm:CORBAException xmi:id="_XRXKIGROEd-Wc9hRqhM12g" base_NamedElement="_XRS4sGROEd-Wc9hRqhM12g" base_DataType="_XRS4sGROEd-Wc9hRqhM12g"/>
...
```

where this exception refers to the following `DataType`:

```xml
      <packagedElement xmi:type="uml:DataType" xmi:id="_XRS4sGROEd-Wc9hRqhM12g" name="InternalError">
...
```

**Comment by eposse** [comment]
Even though the copying of elements described in the previous comment may explain the difference, there is something else that is a bit odd. According to the [EMF Compare Developer Guide](https://www.eclipse.org/emf/compare/documentation/latest/developer/developer-guide.html) and the companion document [Logical Model](https://www.eclipse.org/emf/compare/documentation/latest/developer/logical-model.html), EMF Compare should in theory ignore elements from resources that have not changed.

More precisely, EMF Compare goes through several stages, the first of which, called "Model Resolution", selects and loads all the required resources for the comparison, but only those strictly required, avoiding for example, loading resources which are identical as binaries. The guide also states that `EObject` proxies are not resolved until they are strictly needed. This suggests, that when comparing elements that are proxies to elements in a model library, such as the `setraises` reference of `query` in the snippets above, EMF Compare should not load the model library and the elements should be the same. So, while `Reader` and its attribute `query` may be "different" objects in the model and the library, the `CORBAAttribute` stereotype applied to `query`has a cross reference to the exact same element in the model library, namely the `CORBAException` with id `_XRXKIGROEd-Wc9hRqhM12g`.

Furthermore, as the diff file shows, there is no difference in the `BasicPubSubPapyrus.uml` file at all between the latest revision and the previous revision.

**Comment by eposse** [comment]
The differences are detected in the following functions from EMF Compare:

- `org.eclipse.emf.compare.diff.DefaultDiffEngine.checkForDifferences(Match, Monitor)`
- `org.eclipse.emf.compare.diff.DefaultDiffEngine.computeDifferences(Match, EReference, boolean)`
- `org.eclipse.emf.compare.diff.DefaultDiffEngine.computeSingleValuedReferenceDifferencesThreeWay(Match, EReference)`
- `org.eclipse.emf.compare.diff.DefaultDiffEngine.computeMultiValuedFeatureDifferences(Match, EStructuralFeature, boolean)`
- `org.eclipse.emf.compare.diff.DefaultDiffEngine.createMultiValuedFeatureDifferencesNoOrdering(Match, EStructuralFeature, List<Object>, List<Object>, DifferenceSource)`
- `org.eclipse.emf.compare.utils.IEqualityHelper.matchingValues(Object, Object)`
- `org.eclipse.emf.compare.utils.EqualityHelper.matchingValues(Object, Object)`
- `org.eclipse.emf.compare.utils.EqualityHelper.matchingEObjects(EObject, EObject)`

It's useful to set a conditional breakpoint in line 153 of `org.eclipse.emf.compare.diff.DefaultDiffEngine.checkForDifferences(Match, Monitor)`, with condition `reference.getName().contains("setraises")`.

More precisely, the difference is detected in line 904 of `org.eclipse.emf.compare.diff.DefaultDiffEngine.createMultiValuedFeatureDifferencesNoOrdering(Match, EStructuralFeature, List<Object>, List<Object>, DifferenceSource)`, when the condition `equalityHelper.matchingValues(sideValue, originValue)` is evaluated and returns false unexpectedly.

This `matchingValues` function returns false because it ends up executing line 170:

```java
					equal = matchingEObjects((EObject)object1, (EObject)object2);
```

which in line 248 ends up evaluating

```java
		} else if (getTarget().getMatch(object2) != null || object1.eClass() != object2.eClass()) {
			matching = false;
```
 where 

the two objects have a different `eClass`. The two objects in question are dynamic EObjects:

`org.eclipse.emf.ecore.impl.DynamicEObjectImpl@193184fb (eProxyURI: pathmap://DDS4CCM_DCPS_CONNECTOR_LIBRARIES/DDS_DCPS.uml#_kIfNkGROEd-Wc9hRqhM12g eClass: org.eclipse.emf.ecore.impl.EClassImpl@794b7566 (name: CORBAException) (instanceClassName: null) (abstract: false, interface: false))`

and

`org.eclipse.emf.ecore.impl.DynamicEObjectImpl@18307f07 (eProxyURI: pathmap://DDS4CCM_DCPS_CONNECTOR_LIBRARIES/DDS_DCPS.uml#_kIfNkGROEd-Wc9hRqhM12g eClass: org.eclipse.emf.ecore.impl.EClassImpl@1394943d (name: CORBAException) (instanceClassName: null) (abstract: false, interface: false))`

Note that the `EqualityHelper` has an attribute `equalityHelperExtensionProviderRegistry`, which is used to delegate the matching to an extension perhaps provided via an extension point (?) I haven't found yet what is the extension point, as it is not part of the `org.eclipse.emf.compare` plugin. But this suggests that the EqualityHelper can be configured to deal with a case like the one above.

**Comment by eposse** [comment]
According to Young-Soo, these "copied" elements are actually instantiations of a "template" from the model library, so it is expected to have copies in the model. 

Nevertheless, the difference in the `setraises` attribute does seem unexpected.

**Comment by J17359** [comment]
@eposse can this be closed now that PR #203 has been merged?

**Comment by eposse** [comment]
Yes. Sorry, I forgot.
